File: solutions/blockmatrix.f90
PROGRAM blockmatrix

USE omp_lib
IMPLICIT NONE

INTEGER, PARAMETER :: m = 640, n = 800
DOUBLE PRECISION, ALLOCATABLE :: x(:,:,:)
INTEGER :: i, j, k
DOUBLE PRECISION :: t1, t2, scale

    ALLOCATE(x(m,n,n))

!$omp parallel default(none) private(i,j,k) shared(x)
    do k = 2, n
       do j = 2, n
!$omp do
          do i = 1, m
             x(i,j,k) = 1d0
          end do
!$omp end do nowait
       end do
    end do
!$omp end parallel

    t1 = OMP_GET_WTIME()
    scale = 2.0

!$omp parallel default(none) private(i,j,k) shared(x,scale)
    do k = 2, n
       do j = 2, n
!$omp do
          do i = 1, m
             x(i,j,k) = x(i,j,k-1) + x(i,j-1,k)*scale
          end do
!$omp end do nowait
       end do
    end do
!$omp end parallel

    t2 = OMP_GET_WTIME()

    WRITE(*,*) 'Blockmatrix Time =', t2 - t1, " on ", omp_get_max_threads(), "threads"

END PROGRAM blockmatrix

File: solutions/blockmatrix.c
#include <stdio.h>
#include <omp.h>

#define M 640
#define N 800

double x[N][N][M];

int main(void)
{
    int i, j, k;
    double t1, t2, scale;

#pragma omp parallel default(none) private(i,j,k) shared(x)
    for( k = 1; k < N; k++)
       for( j = 1; j < N; j++)
#pragma omp for
	  for( i = 0; i < M; i++)
             x[k][j][i] = 1.0;

    t1 = omp_get_wtime();
    scale = 2.0;

#pragma omp parallel default(none) private(i,j,k) shared(x,scale)
    for( k = 1; k < N; k++)
       for( j = 1; j < N; j++)
#pragma omp for
	  for( i = 0; i < M; i++)
             x[k][j][i] = x[k-1][j][i] + x[k][j-1][i]*scale;

    t2 = omp_get_wtime();

    printf("%s %d %s %g\n", "Threads = ",  omp_get_max_threads(), "Blockmatrix Time =", t2 - t1);
    return 0;
}
